source/websocket.c

```diff
@@ -1288,35 +1308,112 @@ static int s_decoder_on_payload(struct aws_byte_cursor data, void *user_data) {
     AWS_ASSERT(websocket->thread_data.current_incoming_frame);
     AWS_ASSERT(!websocket->thread_data.is_reading_stopped);
 
-    /* Invoke user cb */
-    bool callback_result = true;
-    if (websocket->on_incoming_frame_payload && !websocket->thread_data.is_midchannel_handler) {
-        size_t window_update_size = data.len;
+    if (websocket->thread_data.is_midchannel_handler) {
+        return s_decoder_on_midchannel_payload(websocket, data);
+    }
 
-        callback_result = websocket->on_incoming_frame_payload(
-            websocket, websocket->thread_data.current_incoming_frame, data, &window_update_size, websocket->user_data);
+    return s_decoder_on_user_payload(websocket, data);
+}
+
+/* Invoke user cb */
+static int s_decoder_on_user_payload(struct aws_websocket *websocket, struct aws_byte_cursor data) {
+    if (!websocket->on_incoming_frame_payload) {
+        return AWS_OP_SUCCESS;
+    }
+
+    size_t window_update_size = data.len;
+
+    if (!websocket->on_incoming_frame_payload(
+            websocket,
+            websocket->thread_data.current_incoming_frame,
+            data,
+            &window_update_size,
+            websocket->user_data)) {
+
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_WEBSOCKET, "id=%p: Incoming payload callback has reported a failure.", (void *)websocket);
```

Comment:
The user callback returns a `bool`, not an `int`, so there might have been an error raised. I made the return type a bool because the error is coming from the user, not from the websocket itself. Usually, callbacks don't have return types at all, but these callbacks needed the ability to immediately stop the websocket from processing more data (I plan to revise the HTTP callbacks in the same way, to have `bool` return values)

The websocket ends up raising AWS_ERROR_HTTP_CALLBACK_FAILURE. Does that all make sense? Is it all too convoluted?

Reply:
Seems reasonable